packages/api/cms-api/src/dependencies/dependencies.service.ts

```diff
@@ -64,17 +77,90 @@ export class DependenciesService {
         console.time("creating block dependency materialized view");
         await this.connection.execute(`DROP MATERIALIZED VIEW IF EXISTS block_index_dependencies`);
         await this.connection.execute(`CREATE MATERIALIZED VIEW block_index_dependencies AS ${viewSql}`);
+        await this.connection.execute(
+            `CREATE UNIQUE INDEX ON block_index_dependencies ("rootId", "rootTableName", "rootColumnName", "blockname", "jsonPath", "targetTableName", "targetId")`,
+        );
         console.timeEnd("creating block dependency materialized view");
 
         console.time("creating block dependency materialized view index");
         await this.connection.execute(`CREATE INDEX block_index_dependencies_targetId ON block_index_dependencies ("targetId")`);
         console.timeEnd("creating block dependency materialized view index");
     }
 
-    async refreshViews(): Promise<void> {
-        console.time("refresh materialized block dependency");
-        await this.connection.execute("REFRESH MATERIALIZED VIEW block_index_dependencies");
-        console.timeEnd("refresh materialized block dependency");
+    async refreshViews(options?: { force?: boolean; awaitRefresh?: boolean }): Promise<void> {
+        const refresh = async (options?: { concurrently: boolean }) => {
+            console.time("refresh materialized block dependency");
+            const blockIndexRefresh = this.refreshRepository.create({ startedAt: new Date() });
+            await this.refreshRepository.getEntityManager().persistAndFlush(blockIndexRefresh);
+
+            await this.connection.execute(`REFRESH MATERIALIZED VIEW ${options?.concurrently ? "CONCURRENTLY" : ""} block_index_dependencies`);
+
+            await this.refreshRepository.getEntityManager().persistAndFlush(Object.assign(blockIndexRefresh, { finishedAt: new Date() }));
+            console.timeEnd("refresh materialized block dependency");
+        };
+
+        const abortActiveRefreshes = async (activeRefreshes: PGStatActivity[]) => {
+            for (const refresh of activeRefreshes) {
+                await this.entityManager.getKnex("write").raw(`SELECT pg_cancel_backend(?);`, [refresh.pid]);
+            }
+        };
+
+        const activeRefreshes: PGStatActivity[] = await this.entityManager
+            .getKnex("read")
+            .select("pid", "state", "query")
+            .from("pg_stat_activity")
+            .where({
+                state: "active",
+            })
+            .andWhereILike("query", "REFRESH MATERIALIZED VIEW%")
+            .andWhereILike("query", "%block_index_dependencies%");
```

Comment:
@nsams FYI: I now used the `pg_stat_activity` table to find the active refreshes. Then I use `pg_cancel_backend()` to cancel active jobs on a forced refresh